File: tutorial/src/flight/src/flight_test.cpp
#include <iostream>
#include <chrono>
#include <vector>
#include <stdint.h>
#include <Eigen/Dense>
#include <Eigen/Geometry>

#include "rclcpp/rclcpp.hpp"

#include "geometry_msgs/msg/point.hpp"

#include "px4_msgs/msg/offboard_control_mode.hpp"
#include "px4_msgs/msg/trajectory_setpoint.hpp"
#include "px4_msgs/msg/vehicle_command.hpp"
#include "px4_msgs/msg/vehicle_odometry.hpp"

using namespace std::chrono;
using namespace std::chrono_literals;
using namespace px4_msgs::msg;

class FlightTest : public rclcpp::Node {
  public:
    FlightTest() : Node("FlightTest") {
      odom_sub_ = this->create_subscription<VehicleOdometry>("/fmu/out/vehicle_odometry", rclcpp::SensorDataQoS(),
                           [this](const VehicleOdometry::SharedPtr msg) {
                           curr_odom_ = *msg; has_odom_ = true;});

      offboard_control_mode_publisher_ = this->create_publisher<OffboardControlMode>("/fmu/in/offboard_control_mode", 10);
      trajectory_setpoint_publisher_ = this->create_publisher<TrajectorySetpoint>("/fmu/in/trajectory_setpoint", 10);
      vehicle_command_publisher_ = this->create_publisher<VehicleCommand>("/fmu/in/vehicle_command", 10);

      auto timer_callback = [this]() -> void {
        if(!has_odom_){
          RCLCPP_WARN(this->get_logger(), "Waiting for odometry...");
          return;
        }

        if (!armed_ && flight_mode_ != LANDED) {
          this->publish_vehicle_command(VehicleCommand::VEHICLE_CMD_DO_SET_MODE, 1, 6);
          this->arm();
        }

        publish_offboard_control_mode();
        publish_trajectory_setpoint();

        offboard_setpoint_counter_++;
      };
      timer_ = this->create_wall_timer(100ms, timer_callback);
    };

  private:
    rclcpp::TimerBase::SharedPtr timer_;
    std::atomic<uint64_t> timestamp_;

    rclcpp::Publisher<OffboardControlMode>::SharedPtr offboard_control_mode_publisher_;
    rclcpp::Publisher<TrajectorySetpoint>::SharedPtr trajectory_setpoint_publisher_;
    rclcpp::Publisher<VehicleCommand>::SharedPtr vehicle_command_publisher_;

    rclcpp::Subscription<VehicleOdometry>::SharedPtr odom_sub_;

    VehicleOdometry curr_odom_;

    enum FlightMode {
        MULTIROTOR = 3,
        FIXED_WING = 4,
        LANDED
    };

    FlightMode flight_mode_ = MULTIROTOR;

    std::vector<std::array<float,3>> waypoints_ = {
      {0.0f, 0.0f, -25.0f},
      {100.0f, 0.0f, -25.0f},
      {200.0f, 100.0f, -25.0f},
      {200.0f, -100.0f, -25.0f},
      {100.0f, 0.0f, -25.0f},
      {0.0f, 0.0f, -25.0f},
      {0.0f, 0.0f, 0.0f}
    };

    uint64_t offboard_setpoint_counter_ {0};
    size_t wp_idx_ {0};

    bool has_odom_ = false;
    bool armed_ = false;

    int hold_counter_ = 0;
    const int HOLD_THRESHOLD = 20;

    void arm();
    void disarm();

    void publish_offboard_control_mode();
    void publish_trajectory_setpoint();
    void publish_vehicle_command(uint16_t command, float param1 = 0.0, float param2 = 0.0);
    void transition(FlightMode mode = MULTIROTOR);

    void publish_vehicle_command(int command, float value);
    float k = 1;
};

void FlightTest::arm() {
  publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0);

  RCLCPP_INFO(this->get_logger(), "Arm command send");
  armed_ = true;
}

void FlightTest::disarm() {
  publish_vehicle_command(VehicleCommand::VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0);

  RCLCPP_INFO(this->get_logger(), "Disarm command send");
  armed_ = false;
}

void FlightTest::publish_offboard_control_mode() {
  OffboardControlMode msg {};
  msg.position = true;
  msg.velocity = flight_mode_ == MULTIROTOR? false:true;
  msg.acceleration = false;
  msg.attitude = false;
  msg.body_rate = false;
  msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
  offboard_control_mode_publisher_->publish(msg);
}

void FlightTest::publish_trajectory_setpoint() {
  if (curr_odom_.timestamp == 0) {
    RCLCPP_WARN(this->get_logger(), "Waiting for odometry...");
    return;
  }

  TrajectorySetpoint msg {};

  auto &wp = waypoints_[wp_idx_];

  Eigen::Vector3f current(curr_odom_.position[0], curr_odom_.position[1], curr_odom_.position[2]);
  Eigen::Vector3f target(wp[0], wp[1], wp[2]);
  Eigen::Vector3f to_wp = target - current;
  float dist_to_wp = to_wp.norm();
  Eigen::Vector3f direction_v(to_wp.x()/dist_to_wp, to_wp.y()/dist_to_wp, to_wp.z()/dist_to_wp);

  if (flight_mode_ == MULTIROTOR && armed_) {
    msg.position = {wp[0], wp[1], wp[2]};

    if (dist_to_wp < 3.0f) {
      hold_counter_++;
      if (hold_counter_ > HOLD_THRESHOLD) {
        RCLCPP_INFO(this->get_logger(), "[MULTIROTOR] Heading to waypoint %ld", wp_idx_ + 1);
        hold_counter_ = 0;
        wp_idx_++;
        if(wp_idx_ == 1) transition(FIXED_WING);
        if(wp_idx_ >= waypoints_.size()) {
          RCLCPP_INFO(this->get_logger(), "[LANDING] Sending land command");
          publish_vehicle_command(VehicleCommand::VEHICLE_CMD_NAV_LAND);
          flight_mode_ = LANDED;
          disarm();
        }
      }
    }
  }

  else if (flight_mode_ == FIXED_WING) {
    msg.position = {wp[0], wp[1], wp[2]};
    msg.velocity = {k*direction_v.x(), k*direction_v.y(), 0};

    if (dist_to_wp < 10.0f) {
      wp_idx_++;
      RCLCPP_INFO(this->get_logger(), "[FIXED_WING] Heading to waypoint %ld", wp_idx_);
    }

    if (wp_idx_ + 2> waypoints_.size()) transition(MULTIROTOR);
  }

  else if (flight_mode_ == LANDED) return;

  msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
  trajectory_setpoint_publisher_->publish(msg);
}

void FlightTest::publish_vehicle_command(uint16_t command, float param1, float param2) {
  VehicleCommand msg {};
  msg.param1 = param1;
  msg.param2 = param2;
  msg.command = command;
  msg.target_system = 1;
  msg.target_component = 1;
  msg.source_system = 1;
  msg.from_external = true;
  msg.timestamp = this->get_clock()->now().nanoseconds() / 1000;
  vehicle_command_publisher_->publish(msg);
}

void FlightTest::transition(FlightMode mode) {
  if (mode == flight_mode_) {
    RCLCPP_INFO(this->get_logger(), "[TRANSITION] Already in desired flight mode, no command sent.");
    return;
  }

  publish_vehicle_command(VehicleCommand::VEHICLE_CMD_DO_VTOL_TRANSITION, static_cast<float>(mode));
  std::string mode_str = (mode == FIXED_WING) ? "Fixed-Wing" : "Multicopter";
  RCLCPP_INFO(this->get_logger(), "[TRANSITION] Sent VTOL transition command: %s", mode_str.c_str());
  flight_mode_ = mode;
}

int main(int argc, char *argv[]) {
  std::cout << "Starting offboard control node..." << std::endl;
  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FlightTest>());

  rclcpp::shutdown();
  return 0;
}

File: ws_sw/flight_test.py
import math 
import sys 

import rclpy 
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy, QoSDurabilityPolicy

from px4_msgs.msg import OffboardControlMode
from px4_msgs.msg import TrajectorySetpoint
from px4_msgs.msg import VehicleStatus
from px4_msgs.msg import VehicleOdometry
from px4_msgs.msg import VehicleCommand
from px4_msgs.msg import VehicleControlMode

class FlightTest(Node):
    def __init__(self):
        super().__init__('flight_test')

        self.offboard_enabled = False
        self.armed = VehicleStatus.ARMING_STATE_DISARMED
        
        self.position = [0.0, 0.0, 0.0]
        self.target = [
            [0.0, 0.0, -10.0],
            [5.0, 0.0, -10.0],
            [5.0, 5.0, -10.0],
            [5.0, -5.0, -10.0],
            [5.0, 0.0, -10.0],
            [0.0, 0.0, -10.0]
        ]

        self.target_index = 0
        self.reached_target_counter = 0
        self.distance_threshold = 2.0
        self.time_threshold = 300 #3초 머무르기 

        qos_profile_pub = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.TRANSIENT_LOCAL,
            history=QoSHistoryPolicy.KEEP_LAST, 
            depth=0
        )

        qos_profile_sub = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=0
        )

        self.subscriber_odometry = self.create_subscription(
            VehicleOdometry, 
            '/fmu/out/vehicle_odometry', 
            self.callback_odometry,
            qos_profile_sub 
        )

        self.subscriber_vehicle_status = self.create_subscription(
            VehicleStatus,
            '/fmu/out/vehicle_status_v1',
            self.callback_vehicle_status,
            qos_profile_sub
        )

        self.subscriber_vehicle_control_mode = self.create_subscription(
            VehicleControlMode,
            '/fmu/out/vehicle_control_mode',
            self.callback_vehicle_control_mode, 
            qos_profile_sub
        )
        
        self.publisher_offboard_mode = self.create_publisher(
            OffboardControlMode,
            '/fmu/in/offboard_control_mode',
            qos_profile_pub
        )

        self.publisher_trajectory_setpoint = self.create_publisher(
            TrajectorySetpoint,
            '/fmu/in/trajectory_setpoint',
            qos_profile_pub
        )

        self.publisher_vehicle_command = self.create_publisher(
            VehicleCommand,
            '/fmu/in/vehicle_command',
            qos_profile_pub
        )

        timer_period = 0.01 
        self.timer = self.create_timer(timer_period, self.callback_cmdloop)
                                                                
    def callback_odometry(self, msg):
        self.position = msg.position  
    
    def callback_vehicle_status(self, msg):
        self.armed = msg.arming_state  

    def callback_vehicle_control_mode(self, msg):
        self.offboard_enabled = msg.flag_control_offboard_enabled 

    def publish_offboard_mode(self):
        msg = OffboardControlMode()
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        msg.position = True
        msg.velocity = False 
        msg.acceleration = False 
        msg.attitude = False 
        msg.body_rate = False 
        msg.thrust_and_torque = False 
        msg.direct_actuator = False 
        self.publisher_offboard_mode.publish(msg)

    def publish_vehicle_command(self, command, param1, param2):
        msg = VehicleCommand()
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        msg.param1 = float(param1)
        msg.param2 = float(param2)
        msg.command = command 
        msg.from_external = True
        self.publisher_vehicle_command.publish(msg)

    def publish_arm_command(self):
        self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 1.0, 0.0)
        print("Arm command send") 
    
    def publish_disarm_command(self):
        self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_COMPONENT_ARM_DISARM, 0.0, 0.0)
        print("Disarm command send") 
    
    def publish_trajectory_setpoint(self, p):
        msg = TrajectorySetpoint()
        msg.timestamp = int(self.get_clock().now().nanoseconds / 1000)
        msg.position = p 
        self.publisher_trajectory_setpoint.publish(msg)
    
    def close_enough(self):
        distance_to_target = math.dist(self.position, self.target[self.target_index])
        if distance_to_target < self.distance_threshold:
            return True
        else:
            return False
        
    def callback_cmdloop(self):
        if self.armed == VehicleStatus.ARMING_STATE_DISARMED:
            self.publish_arm_command()
            self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_DO_SET_MODE, 1.0, 6.0)

        self.publish_offboard_mode()

        self.publish_trajectory_setpoint(self.target[self.target_index])

        if self.close_enough():
            self.reached_target_counter += 1.0
            if self.reached_target_counter > self.time_threshold:
                if self.target_index < len(self.target) - 1:
                    self.target_index += 1 
                    self.reached_target_counter = 0.0
                else: #land 추가 
                    self.publish_vehicle_command(VehicleCommand.VEHICLE_CMD_NAV_LAND, 0.0, 0.0)
                    self.get_logger().info("LAND COMMAND SEND")
                    self.timer.cancel()
                    sys.exit(0)
                
def main(args=None):
    rclpy.init(args=args)
    flight_test = FlightTest()
    rclpy.spin(flight_test)
    flight_test.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
